src/Data/Functor/Foldable/TH.hs

```diff
@@ -229,17 +229,17 @@ makePrimForDI rules mkInstance'
                 , datatypeInstTypes = instTys
                 , datatypeCons      = cons
                 , datatypeVariant   = variant }) = do
+#if MIN_VERSION_th_abstraction(0,5,0)
+    when (variant == TH.Abs.TypeData) $
+      fail "makeBaseFunctor: `type data` declarations are not supported."
+#endif
     when isDataFamInstance $
       fail "makeBaseFunctor: Data families are currently not supported."
     makePrimForDI' rules mkInstance'
                    (variant == Newtype) tyName
                    (map toTyVarBndr instTys) cons
   where
-    isDataFamInstance = case variant of
-                          DataInstance    -> True
-                          NewtypeInstance -> True
-                          Datatype        -> False
-                          Newtype         -> False
+    isDataFamInstance = variant `elem` [DataInstance, NewtypeInstance]
```

Comment:
I made this change to avoid `-Wincomplete-patterns` warnings involving `TypeData`. Another way to fix this would be to add an explicit case on `TypeData`, but that would require more CPP. Let me know which way you would prefer.

Reply:
> Therefore, how about an allow-list instead of a deny-list?

I'd be OK with that as well. The error message might need to be cleaned up a bit, since might be confused to see an error message mentioning things like `DatatypeInstance`, which is an implementation detail that leaks through from `th-abstraction`. I think I can make the error messages (relatively) nice and combine everything into a single pattern-match based on your suggestion.

Reply:
I've pushed a new version of the code that combines everything into a `checkAllowed` function, which is the sole place where I introduce CPP.